raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -158,6 +167,8 @@ public KafkaRaftClient(NetworkChannel channel,
 
     private void applyCommittedRecordsToStateMachine() {
         quorum.highWatermark().ifPresent(highWatermark -> {
+            log.updateHighWatermark(highWatermark);
```

Comment:
This is sort of a meta question, have we considered letting the ReplicatedLog to maintain the high watermark? From what I could see, we need to call `log.updateHighWatermark` separately from updating the quorum high watermark, which is a bit easy to get lost.

Reply:
Yeah, there is not a clear separation of responsibility when it comes to the high watermark. Ideally I would prefer if the `Log` implementation was not aware of the high watermark at all, but it is convenient for reasons outside of the Raft implementation for it to be tracked there. In fact, the only reason we need this line at all is because `Log.read` only allows us to specify the isolation of the read rather than the limit offset as the `ReplicatedLog` API suggests. Let me file a separate jira for this and we can try to come up with a better approach.